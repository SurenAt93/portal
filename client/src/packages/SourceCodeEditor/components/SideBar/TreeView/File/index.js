import React from 'react';
import PropTypes from 'prop-types';

import Button from '@material-ui/core/Button';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';

import classNames from 'classnames';

// Icons
import FileIcon from '@material-ui/icons/Notes';

const File = ({ name, handleFileOpen, onContextMenu, open, path }) => (
  <ListItem className="files__item file">
    <div className="files__summary">
      <ListItemIcon><FileIcon /></ListItemIcon>
      <Button
        onContextMenu={onContextMenu}
        onClick={_ => handleFileOpen(path)}
        className={classNames({ 'files__file--open': open })}
      >
        <ListItemText primary={name}/>
      </Button>
    </div>
  </ListItem>
);

File.propTypes = {
  name: PropTypes.string.isRequired,
  path: PropTypes.string.isRequired,
  open: PropTypes.bool,
  handleFileOpen: PropTypes.func,
  onContextMenu: PropTypes.func.isRequired,
};

File.defaultProps = {
  handleFileOpen: _ => {},
  open: false,
};

export default File;
