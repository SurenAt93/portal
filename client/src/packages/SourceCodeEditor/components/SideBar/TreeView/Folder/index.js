import React, { PureComponent, Fragment } from 'react';
import PropTypes from 'prop-types';

import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import ListItemIcon from '@material-ui/core/ListItemIcon';

// Icons
import FolderIcon from '@material-ui/icons/Folder';
import FolderIconOpen from '@material-ui/icons/FolderOpen';

class Folder extends PureComponent {
  state = { open: false };

  handleClick = _ => this.setState(state => ({ open: !state.open }));

  render() {
    const { name, path, children, draw } = this.props;

    return (
      <Fragment>
        <ListItem
          button
          onContextMenu={this.props.onContextMenu}
          onClick={this.handleClick} className="files__item"
        >
          <div className="files__summary">
            <ListItemIcon>
              {
                this.state.open
                  ? <FolderIconOpen />
                  : <FolderIcon />
              }
            </ListItemIcon>
            <ListItemText primary={name}/>
          </div>
        </ListItem>
        {
          this.state.open && <div className="collapse">
            <List className="files__folder" dense={true}>
              {children.map(child => draw(child, path))}
            </List>
          </div>
        }
      </Fragment>
    );
  }
}

Folder.propTypes = {
  name: PropTypes.string.isRequired,
  path: PropTypes.string.isRequired,
  children: PropTypes.array.isRequired,
  draw: PropTypes.func.isRequired,
  onContextMenu: PropTypes.func.isRequired,
};

export default Folder;

