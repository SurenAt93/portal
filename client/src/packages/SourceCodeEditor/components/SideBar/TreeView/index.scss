.files {
  width: 100%;
  height: 100%;
  overflow: scroll;
  white-space: nowrap;

  &__file {
    &--open {
      border-right: 4px solid #65cdc4;
      border-top-right-radius: 0;
      border-bottom-right-radius: 0;
    }
  }

  .file {
    button {
      padding-left: 10px;
    }

    svg {
      margin-right: 0;
    }
  }

  ul {
    height: 100%;
    min-width: max-content;

    li {
      padding-top: 4px;
      padding-bottom: 4px;
      padding-right: 0;
    }
  }

  &__item {
    flex-direction: column;
    align-items: flex-start;

    * {
      margin: 0;
      padding: 0;
    }
  }

  &__folder {
    margin-left: 15px;
  }

  &__summary {
    display: flex;
    align-items: center;
    width: 100%;

    > svg {
      margin-right: 10px;
    }

    button {
      text-transform: none;
      width: 100%;
      text-align: start;
    }
  }
}