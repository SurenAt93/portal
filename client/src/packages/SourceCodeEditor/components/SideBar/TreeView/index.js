import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import List from '@material-ui/core/List';

// Components
import File from './File';
import Folder from './Folder';

// Styles
import './index.scss';

class TreeView extends PureComponent {
  draw = (
    { name, children, open },
    parent,
    path = parent ? `${parent}${name}${children ? '/' : ''}` : name,
  ) => children
    ? <Folder
      key={path}
      path={path}
      name={name}
      open={open}
      children={children}
      draw={this.draw}
      onContextMenu={this.props.onFolderContextMenu}
      onToggle={this.props.onFolderToggle}
    />
    : <File
      onOpen={this.props.onFileOpen}
      key={path}
      path={path}
      name={name}
      open={this.props.openFilePath === path}
      onContextMenu={this.props.onFileContextMenu}
    />;

  render() {
    const { data } = this.props;

    return (
      <div className="files">
        {
          data
            ? (
              // TODO ::: This list should be virtualized
              <List dense={true}>
                {this.draw(data)}
              </List>
            )
            : 'Empty...'
        }
      </div>
    );
  }
}

TreeView.propTypes = {
  data: PropTypes.object,
  openFilePath: PropTypes.string,
  onFileOpen: PropTypes.func,
  onFolderToggle: PropTypes.func,
  onFolderContextMenu: PropTypes.func.isRequired,
  onFileContextMenu: PropTypes.func.isRequired,
};

TreeView.defaultProps = {
  handleFileOpen: _ => {},
};

export default TreeView;
