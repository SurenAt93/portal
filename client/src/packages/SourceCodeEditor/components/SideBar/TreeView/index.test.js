import React from 'react';
import TreeView from '.';

describe('<TreeView />', () => {
  it('should render correctly', () => {
    const component = shallow(
      <TreeView
        onFolderContextMenu={_ => {}}
        onFileContextMenu={_ => {}}
      />
    );

    expect(component).toMatchSnapshot();
  });
});
