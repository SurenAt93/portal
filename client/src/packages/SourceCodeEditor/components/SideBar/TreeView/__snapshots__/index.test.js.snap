// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TreeView /> should render correctly 1`] = `
<div
  className="files"
>
  Empty...
</div>
`;
